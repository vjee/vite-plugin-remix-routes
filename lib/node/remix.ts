import { defineConventionalRoutes } from "@remix-run/dev/dist/config/routesConvention";
import {
  ConfigRoute,
  RouteManifest,
  defineRoutes,
} from "@remix-run/dev/dist/config/routes";
import type { PluginOptions } from "./index";
import type { AppConfig } from "@remix-run/dev/dist/config";
import type { RequireOnly } from "./utils";

export type RemixOptions = Pick<
  AppConfig,
  "appDirectory" | "routes" | "ignoredRouteFiles"
>;

type GetRouteOptions = Omit<PluginOptions, "importModule"> &
  RequireOnly<RemixOptions, "appDirectory">;

/**
 * See `readConfig` in @remix-run/dev/config.ts
 */
export async function getRoutes(options: GetRouteOptions) {
  const {
    appDirectory,
    is404Route = (route) => route.id.endsWith("/404"),
    ignoredRouteFiles,
    routes,
  } = options;

  const routeManifest: RouteManifest = {
    root: { path: "", id: "root", file: "routes/index" },
  };

  const conventionalRoutes = defineConventionalRoutes(
    appDirectory,
    ignoredRouteFiles
  );

  for (const key of Object.keys(conventionalRoutes)) {
    const route = conventionalRoutes[key];
    routeManifest[route.id] = {
      ...route,
      parentId: route.parentId || "root",
    };
  }

  if (routes) {
    let manualRoutes = await routes(defineRoutes);
    for (const key of Object.keys(manualRoutes)) {
      const route = manualRoutes[key];
      routeManifest[route.id] = {
        ...route,
        parentId: route.parentId || "root",
      };
    }
  }

  const routeConfig = createRoutes(routeManifest)[0].children;

  // This is not part of remix.
  const modifyRoute = (route: Route): Route => ({
    ...route,
    path: is404Route(route) ? "*" : route.path,
    children: route.children.map(modifyRoute),
  });

  return routeConfig.map(modifyRoute);
}

/**
 * See `createClientRoutes` in @remix-run/react/routes.tsx
 */
export function createRoutes(
  routeManifest: RouteManifest,
  parentId?: string
): Route[] {
  return Object.keys(routeManifest)
    .filter((key) => routeManifest[key].parentId === parentId)
    .map((key) => {
      const route = createRoute(routeManifest[key]);
      route.children = createRoutes(routeManifest, route.id);
      return route;
    });
}

/**
 * See `createClientRoute` in @remix-run/react/routes.tsx
 */
export function createRoute(route: ConfigRoute): Route {
  return {
    id: route.id,
    file: route.file,

    path: route.path || "",
    index: !!route.index,
    children: [],
  };
}

export interface Route {
  // custom properties
  id: string;
  file: string;

  // react-router route properties
  path: string;
  index: boolean;
  children: Route[];
}
