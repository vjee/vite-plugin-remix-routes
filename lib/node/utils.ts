import type { Route } from "./remix";

export type RequireOnly<Object, Keys extends keyof Object> = Omit<
  Object,
  Keys
> &
  Required<Pick<Object, Keys>>;

export interface Context {
  prefix: string;
  importMode?: (route: Route) => "sync" | "async";
}

interface Components {
  sync: string[];
  async: string[];
}

export function stringifyRoutes(routes: Route[], context: Context) {
  const components: Components = { sync: [], async: [] };
  const routesString = routesToString(routes, context, components);

  return {
    routesString,
    componentsString: [...components.sync, ...components.async].join("\n"),
  };
}

function routesToString(
  routes: Route[],
  context: Context,
  components: Components
) {
  return (
    "[" +
    routes.map((route) => routeToString(route, context, components)).join(",") +
    "]"
  );
}

function routeToString(
  route: Route,
  context: Context,
  components: Components
): string {
  const componentName = getRouteComponentName(route);
  const componentPath = `${context.prefix}/${route.file}`;
  const importMode = context.importMode?.(route) || "sync";

  const props = new Map<string, string>();

  if (route.path !== "") {
    props.set("path", `'${route.path}'`);
  }

  if (importMode === "async") {
    components.async.push(
      `const ${componentName} = () => import('${componentPath}');`
    );

    props.set("element", `createElement(lazy(${componentName}))`);
    props.set("loader", componentName);
  }

  if (importMode === "sync") {
    components.sync.push(`import ${componentName} from '${componentPath}';`);

    props.set("element", `createElement(${componentName})`);
  }

  if (route.index === true) {
    props.set("index", "true");
  }

  if (route.children.length) {
    const children = routesToString(route.children, context, components);
    props.set("children", children);
  }

  return (
    "{" + [...props.entries()].map(([k, v]) => `${k}:${v}`).join(",") + "}"
  );
}

function getRouteComponentName(route: Route) {
  return route.id
    .split(/[/.]/)
    .map((str) => str.replace(/^\w/, (c) => c.toUpperCase()))
    .join("");
}
