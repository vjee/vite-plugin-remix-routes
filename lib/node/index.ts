import path from "node:path";
import fs from "node:fs";

import type { Plugin } from "vite";

import { getRoutes, RemixOptions } from "./remix";
import type { Route } from "./remix";
import { stringifyRoutes } from "./utils";

export interface Options extends PluginOptions, RemixOptions {}
export interface PluginOptions {
  /**
   * An absolute path to the folder containing the `routes` folder.
   * This will most likely be your `/src` folder.
   *
   * @deprecated Use `appDirectory` instead
   */
  appDir?: string;

  /**
   * A function that receives a `Route` to determine if the route's component
   * should be imported synchronously or asynchronously.
   *
   * @default () => "sync"
   */
  importMode?: (route: Route) => "async" | "sync";

  /**
   * A function that receives a `Route` to determine if it should be a 404 route. (`path="*"`)
   * By default this matches the same 404 file as Remix does.
   * Keep in mind this only receives top level routes, so you can't mark nested routes
   * as 404 routes.
   *
   * @default (route) => route.id === "routes/404"
   */
  is404Route?: (route: Route) => boolean;
}

function plugin(options: Options = {}): Plugin {
  const virtualModuleId = "virtual:remix-routes";

  const {
    appDir: _appDir,
    appDirectory = _appDir,
    importMode,
    ...otherOptions
  } = options;

  let dir = "";

  // Check both src for backwards compatibility and app for remix's normal default
  if (appDirectory) {
    dir = path.resolve(process.cwd(), appDirectory);
  } else {
    const srcDir = path.resolve(process.cwd(), "src");
    const appDir = path.resolve(process.cwd(), "app");

    if (!dir && fs.existsSync(srcDir)) {
      dir = srcDir;
    }

    if (!dir && fs.existsSync(appDir)) {
      dir = appDir;
    }

    if (!dir) {
      throw new Error(
        "[vite-plugin-remix-routes] appDirectory not found, please specify it in the plugin's config"
      );
    }
  }

  const prefix = `.${path.sep}${path.relative(process.cwd(), dir)}`;

  return {
    name: "vite-plugin-remix-routes",

    resolveId(id) {
      if (id === virtualModuleId) {
        return id;
      }
    },

    async load(id) {
      if (id === virtualModuleId) {
        const routes = await getRoutes({ appDirectory: dir, ...otherOptions });

        const { routesString, componentsString } = stringifyRoutes(routes, {
          prefix,
          importMode,
        });

        return (
          `import { createElement, lazy, useEffect } from 'react';\n` +
          `${componentsString}\n` +
          `export default ${routesString};\n`
        );
      }
    },
  };
}

export default plugin;
export { getRoutes, stringifyRoutes };

export type { Route };
