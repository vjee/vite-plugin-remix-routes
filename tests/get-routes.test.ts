import path from "node:path";
import url from "node:url";

import { getRoutes } from "../lib/node";

const dirname = path.dirname(url.fileURLToPath(import.meta.url));
const appDirectory = path.join(dirname, "../examples/basic/src");

test("getRoutes", async () => {
  const routes = await getRoutes({ appDirectory });

  expect(routes).toMatchSnapshot();
});

test("getRoutes with manual routes specified by options", async () => {
  const file = "routes/demos/about/whoa.tsx";
  const path = "/foo/bar";
  const routes = await getRoutes({
    appDirectory,
    routes: async (defineRoutes) => {
      return defineRoutes((route) => {
        route(path, file);
      });
    },
  });

  expect(routes).toMatchSnapshot();
  expect(routes.find((r) => r.path === path)).toContain({ file });
});

test("getRoutes ignores files from ignoredRouteFiles", async () => {
  // Verify the positive case first
  let routes = await getRoutes({
    appDirectory,
  });
  const testRoute = routes.find((r) => r.file === "routes/one.two.three.tsx");
  expect(routes).toContain(testRoute);

  // Then ensure that it actually is removed
  routes = await getRoutes({
    appDirectory,
    ignoredRouteFiles: ["one.two.three.tsx"],
  });
  expect(routes).not.toContain(testRoute);
});
