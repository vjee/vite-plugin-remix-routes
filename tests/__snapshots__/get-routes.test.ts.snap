// Vitest Snapshot v1

exports[`getRoutes 1`] = `
[
  {
    "children": [],
    "file": "routes/one.two.three.tsx",
    "id": "routes/one.two.three",
    "index": false,
    "path": "one/two/three",
  },
  {
    "children": [
      {
        "children": [],
        "file": "routes/demos/params/index.tsx",
        "id": "routes/demos/params/index",
        "index": true,
        "path": "",
      },
      {
        "children": [],
        "file": "routes/demos/params/$id.tsx",
        "id": "routes/demos/params/$id",
        "index": false,
        "path": ":id",
      },
    ],
    "file": "routes/demos/params.tsx",
    "id": "routes/demos/params",
    "index": false,
    "path": "demos/params",
  },
  {
    "children": [
      {
        "children": [],
        "file": "routes/demos/about/index.tsx",
        "id": "routes/demos/about/index",
        "index": true,
        "path": "",
      },
      {
        "children": [],
        "file": "routes/demos/about/whoa.tsx",
        "id": "routes/demos/about/whoa",
        "index": false,
        "path": "whoa",
      },
      {
        "children": [],
        "file": "routes/demos/about/404.tsx",
        "id": "routes/demos/about/404",
        "index": false,
        "path": "*",
      },
    ],
    "file": "routes/demos/about.tsx",
    "id": "routes/demos/about",
    "index": false,
    "path": "demos/about",
  },
  {
    "children": [],
    "file": "routes/index.tsx",
    "id": "routes/index",
    "index": true,
    "path": "",
  },
  {
    "children": [],
    "file": "routes/404.tsx",
    "id": "routes/404",
    "index": false,
    "path": "*",
  },
]
`;

exports[`getRoutes with manual routes specified by options 1`] = `
[
  {
    "children": [],
    "file": "routes/one.two.three.tsx",
    "id": "routes/one.two.three",
    "index": false,
    "path": "one/two/three",
  },
  {
    "children": [
      {
        "children": [],
        "file": "routes/demos/params/index.tsx",
        "id": "routes/demos/params/index",
        "index": true,
        "path": "",
      },
      {
        "children": [],
        "file": "routes/demos/params/$id.tsx",
        "id": "routes/demos/params/$id",
        "index": false,
        "path": ":id",
      },
    ],
    "file": "routes/demos/params.tsx",
    "id": "routes/demos/params",
    "index": false,
    "path": "demos/params",
  },
  {
    "children": [
      {
        "children": [],
        "file": "routes/demos/about/index.tsx",
        "id": "routes/demos/about/index",
        "index": true,
        "path": "",
      },
      {
        "children": [],
        "file": "routes/demos/about/404.tsx",
        "id": "routes/demos/about/404",
        "index": false,
        "path": "*",
      },
    ],
    "file": "routes/demos/about.tsx",
    "id": "routes/demos/about",
    "index": false,
    "path": "demos/about",
  },
  {
    "children": [],
    "file": "routes/demos/about/whoa.tsx",
    "id": "routes/demos/about/whoa",
    "index": false,
    "path": "/foo/bar",
  },
  {
    "children": [],
    "file": "routes/index.tsx",
    "id": "routes/index",
    "index": true,
    "path": "",
  },
  {
    "children": [],
    "file": "routes/404.tsx",
    "id": "routes/404",
    "index": false,
    "path": "*",
  },
]
`;
